### PR TITLE
Hci initialization sample

### PR DESCRIPTION
**Pull request by kolodgrz**
This is sample initial initialization for testing puroposes.

### CODE REVIEW

**Review by bgix — CHANGES_REQUESTED** [review]
Hi Lukasz,

Many (Most?) HCI controllers have a limited amount of command space, and while all of your commands *should* be able to be done synchronously, there is a chance that not all controllers will successfully execute them all.

The correct method then, is to create a command chain, and call bt_hci_chain once per callback, and this will cause the next command to be issued only after the prior commands Command Complete has arrived.

Also:  I don't remember if bt_hci_register does anything of consequence over the HCI interface, but if it does, it should probably be called at the end of the command chain after the other commands have been issued.

for an example of a command chain, see in that same file, set_send_adv_data() where its first step is setting the Advertising data (BT_HCI_CMD_LE_SET_ADV_DATA), and it's next step in the command chain is  calling enabling advertising (BT_HCI_CMD_LE_SET_ADV_ENABLE).  This should ensure that it works with the widest range of controllers as possible.

File: mesh/mesh-io-generic.c [hunk 1]
```diff
@@ -177,6 +177,66 @@ static void event_callback(const void *buf, uint8_t size, void *user_data)
...
 }
 
+static void configure_hci(struct mesh_io_private *io)
+{
+	struct bt_hci_cmd_le_set_scan_parameters cmd;
+	struct bt_hci_cmd_set_event_mask cmd_sem;
+	struct bt_hci_cmd_le_set_event_mask cmd_slem;
+
+	/* Set scan parameters */
+	cmd.type=0x01; /* Active scanning. Scanning PDUs may be sent. */
```
Comment (jhedberg) [inline mesh/mesh-io-generic.c]: Should be passive scanning, not active scanning

File: mesh/mesh-io-generic.c [hunk 2]
```diff
@@ -177,6 +177,66 @@ static void event_callback(const void *buf, uint8_t size, void *user_data)
...
+	cmd.interval=0x0010; /* Scan Interval = N * 0.625ms */
+	cmd.window=0x0010; /* Scan Window = N * 0.625ms */
```
Comment (jhedberg) [inline mesh/mesh-io-generic.c]: Based on recent experimentation you actually get much better performance with 30ms scan interval and window (at least with Zephyr controller), so I'd recommend changing to that.

**Review by jhedberg — CHANGES_REQUESTED** [review]

**Comment by rymanluk** [comment]
@bgix, agree with you and afaik most of controllers has 1 ticket for commands. Actually what you just said is a main comment I had to this patch and that is why @kolodgrz did not send to upstream yet. He use it as a temporary quick fix.
Anyway, I think it should be quite easy to fix it as `bt_send_hci` has response callback, so basically each command should be send in the response calllback of previous one and it should work

File: mesh/mesh-io-generic.c (hunk 1 above)
Comment (kolodgrz) [reply mesh/mesh-io-generic.c]: Agree.

File: mesh/mesh-io-generic.c (hunk 2 above)
Comment (kolodgrz) [reply mesh/mesh-io-generic.c]: Ok, will change.

**Comment by kolodgrz** [comment]
> @bgix, agree with you and afaik most of controllers has 1 ticket for commands. Actually what you just said is a main comment I had to this patch and that is why @kolodgrz did not send to upstream yet. He use it as a temporary quick fix.
> Anyway, I think it should be quite easy to fix it as `bt_send_hci` has response callback, so basically each command should be send in the response calllback of previous one and it should work

Yes :) I'll update patch. This was quick fix (works for me hci_uart application on nrf)

**Comment by jhedberg** [comment]
@kolodgrz @bgix btw, we tried the patch as well with a Zephyr controller (hci_uart on nRF51 USB dongle) which always gives ncmd=1, and it worked fine - so it doesn't seem like the asynchronous callback chaining is strictly needed? (perhaps only if you want to catch intermediate errors?)

**Comment by jhedberg** [comment]
I should add, that the asynchronous callback chaining approach is quite awful (regarding code readability) for anything more than a couple of HCI commands, so something like a 10+ command HCI init procedure should IMO use some other approach. If an appropriate HCI API doesn't exist then that needs to be added. Note: this is not really relevant to the current task at hand, i.e. getting something working this week for Junction.

**Comment by kolodgrz** [comment]
Clean style and merge request to master branch required. Closing

**Pull request closed without merging**